lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,472 @@
+	if txnSizeBytes != 0 && bav.GlobalParamsEntry.MinimumNetworkFeeNanosPerKB != 0 {
+		// Make sure there isn't overflow in the fee.
+		if txn.TxnFeeNanos != ((txn.TxnFeeNanos * 1000) / 1000) {
+			return nil, 0, 0, 0, RuleErrorOverflowDetectedInFeeRateCalculation
+		}
+		// If the fee is less than the minimum network fee per KB, return an error.
+		if (txn.TxnFeeNanos*1000)/uint64(txnSizeBytes) < bav.GlobalParamsEntry.MinimumNetworkFeeNanosPerKB {
+			return nil, 0, 0, 0, RuleErrorTxnFeeBelowNetworkMinimum
+		}
+	}
+
+	// Construct a UtxoOp for the atomic transactions wrapper.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:                   OperationTypeAtomicTxnsWrapper,
+		AtomicTxnsInnerUtxoOps: innerUtxoOps,
+	})
+
+	return utxoOpsForTxn, totalInput, totalOutput, totalFees, nil
+}
+
+func _verifyAtomicTxnsWrapper(txn *MsgDeSoTxn) error {
+	// Since the wrapper does not require a public key nor a corresponding signature, we force both
+	// the transaction public key to be the ZeroPublicKey and the signature to be nil.
+	if !NewPublicKey(txn.PublicKey).IsZeroPublicKey() {
+		return RuleErrorAtomicTxnsWrapperPublicKeyMustBeZero
+	}
+	if txn.Signature.Sign != nil {
+		return RuleErrorAtomicTxnsWrapperSignatureMustBeNil
+	}
+
+	// We force TxInputs on the wrapper to be empty for several reasons:
+	//	(1) This is consistent with the logic found in _connectBasicTransferWithExtraSpend()
+	//		that forces TxInputs to be empty following the balance model fork.
+	//	(2) Allowing TxInputs to not be empty would lead to an attack vector were the transaction
+	//		size may be bloated with random TxInputs that do nothing.
+	//	(3) Leads to consistent hashing for the same atomic transaction wrapper and its inner transactions.
+	//	(4) It's generally safer to be more restrictive on the transaction structure.
+	if len(txn.TxInputs) != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroInputs
+	}
+
+	// We force TxOutputs on the wrapper to be empty even though this field is still used post balance model fork.
+	// The reason is this transaction is effectively "signed" by the ZeroPublicKey which is a potential
+	// burn address. Hence, allowing TxOutputs to be populated would enable un-burning DESO which we do not want.
+	if len(txn.TxOutputs) != 0 {
+		return RuleErrorAtomicTxnsWrapperMustHaveZeroOutputs
+	}
+
+	// There exists three design options for txn.TxnFeeNanos rules in atomic transaction wrappers:
+	//	(1) Force txn.TxnFeeNanos to equal zero.
+	//	(2) Force txn.TxnFeeNanos to equal the sum of the internal transaction's txn.TxnFeeNanos fields.
+	//	(3) Ignore txn.TxnFeeNanos entirely.
+	//
+	// Because txn.TxnFeeNanos gets used in several places for non-connection logic (e.g. BMF),
+	// it's important to use design option (2) to be consistent across core. This check as a result
+	// becomes extremely important in _connectAtomicTxnsWrapper().
+	var totalInnerTxnFees uint64
+	var err error
```

Comment:
Should we add a validation that the number of inner txns is non-zero? seems like a decent sanity check to me.